common/lib/capa/capa/capa_problem.py

```diff
@@ -492,6 +495,121 @@ def get_answer_ids(self):
             answer_ids.append(results.keys())
         return answer_ids
 
+    def find_question_label(self, answer_id):
+        """
+        Obtain the most relevant question text for a particular answer.
+
+        E.g. in a problem like "How much is 2+2?" "Two"/"Three"/"More than three",
+        this function returns the "How much is 2+2?" text.
+
+        It uses, in order:
+        - the question prompt, if the question has one
+        - the <p> or <label> element which precedes the choices (skipping descriptive elements)
+        - a text like "Question 5" if no other name could be found
+
+        Arguments::
+            answer_id: a string like "98e6a8e915904d5389821a94e48babcf_13_1"
+
+        Returns:
+            a string with the question text
+        """
+        # Some questions define a prompt with this format:   >>This is a prompt<<
+        assert answer_id in self.problem_data
+        prompt = self.problem_data[answer_id].get('label')
+
+        if prompt:
+            question_text = prompt.striptags()
+        else:
+            # If no prompt, then we must look for something resembling a question ourselves
+            #
+            # We have a structure like:
+            #
+            # <p />
+            # <optionresponse id="a0effb954cca4759994f1ac9e9434bf4_2">
+            #   <optioninput id="a0effb954cca4759994f1ac9e9434bf4_3_1" />
+            # <optionresponse>
+            #
+            # Starting from  answer (the optioninput in this example) we go up and backwards
+            xml_elems = self.tree.xpath('//*[@id="' + answer_id + '"]')
+            assert len(xml_elems) == 1
+            xml_elem = xml_elems[0].getparent()
+
+            # Get the element that probably contains the question text
+            questiontext_elem = xml_elem.getprevious()
+
+            # Go backwards looking for a <p> or <label>, but skip <description> because it doesn't
+            # contain the question text.
+            #
+            # E.g if we have this:
+            #   <p /> <description /> <optionresponse /> <optionresponse />
+            #
+            # then from the first optionresponse we'll end with the <p>.
+            # If we start in the second optionresponse, we'll find another response in the way,
+            # stop early, and instead of a question we'll report "Question 2".
+            SKIP_ELEMS = ['description']
+            LABEL_ELEMS = ['p', 'label']
+            while questiontext_elem is not None and questiontext_elem.tag in SKIP_ELEMS:
+                questiontext_elem = questiontext_elem.getprevious()
+
+            if questiontext_elem is not None and questiontext_elem.tag in LABEL_ELEMS:
+                question_text = questiontext_elem.text
+            else:
+                # For instance 'd2e35c1d294b4ba0b3b1048615605d2a_2_1' contains 2,
+                # which is used in question number 1
+                question_nr = int(answer_id.split('_')[-2]) - 1
```

Comment:
First question is 1 and the answer counter [rapidly gets incremented again|https://github.com/edx/edx-platform/blob/clemente%2Fgenerate-report-data--for-capa-problems/common/lib/capa/capa/capa_problem.py#L1010] and becomes 2. So for the first question,  2-1=1. I added an assert to say (in code) that it's never 0.

Reply:
I don't think there should be any asserts in the final code at all. It's better to end up with a "Question 0" or even "Question -1" than to have the entire report fail because at one point there is an assertion that failed.

Reply:
I'm not sure what's better (CI failure or user seeing different results). I have found many errors thanks to asserts, and I may be overusing them. Anytime I want write a comment like `# question_id is never 0` I rewrite as `assert` for Python to enforce it.
If it ever fails it means that the assumptions are wrong, and the code would need to be checked. It wouldn't be the user who sees the error, but the CI (and the programmer who changed the assumptions).
If you prefer I'll leave the „is never 0“ as comment.

Reply:
I might be wrong, but I don't think we are running the LMS with asserts disabled. In which case the end-user will see it.
If for some odd case the assert fails, then the report generation will just fail, right? Cause I am not catching assertion errors. Wherever you expect there might be an assertion error in some odd case, you can just replace ```assert x y`` with ``if x!=y: raise NotImplementedError``, because while the assumption in the ``generate_report_data`` code might be wrong, for the end user it's more useful to get a less readable report than no report at all.

Reply:
asserts are just for things that are always true, so it can't be transformed to an `if`. We don't need a different behavior for when the implementation is right and when isn't; it must always be right. This can be checked in a test case, but with `assert` it's easier and it serves as documentation.
„Better less readable report or no report“: depends on how correct must the product be.
I removed the assert and left the comment. It shouldn't ever happen anyway, the capa implementation should always work (until it doesn't).

Reply:
There is already a working way to get user state without the ``generate_report_data`` method. The issue is that it's not easily readable. So if there are edge cases where this method doesn't work yet, it makes sense to raise the NotImplementError so I can fall back to the old way. So the product is correct either way. 

I understand what you're saying that the asserts will always be true. However, edge cases are always unexpected. For instance, what happens when a problem in a live course is edited and the answer_ids change but users still have responses with old ids? Is it possible that an assert will fail in that case? I don't know honestly, it'll require diving into the code a bit to find out. 

Currently some of the asserts are also making testing harder for me. For example: ``assert answer_id in lcp.problem_data`` in ``generate_report_data`` is forcing me to mock the problem_data when it's never even used in the code.